kb/2022-03-10/multiple_nics_vm_connectivity.md

```diff
@@ -0,0 +1,140 @@
+
+Mostly Linux distribution will treat multiple NICs without any dependency and allow only one default route with the same distance by default. It will start multiple DHCP clients simultaneously. If we have two NICs with DHCP and both have gateway IP in DHCP response, two DHCP clients will race for setting default route to Linux network. According to different implementations of DHCP script, the former and latter will not always win to set up the default route. This will cause an unstable default route in every DHCP renewing process or machine boot up. VM's connection will be unstable if the default route is changing.
+
+## How to avoid random connectivity issue
+
+We could avoid this issue easily by keeping VM only one NICs to serve network. We keep only one IP and one gateway in your system.
+
+But not always we could use only one NIC to connect. For example, our infrasturcture will have storage network and service network. For security reason, storage network will be isolated from service network and have an independent subnet. In this situation, we must have two NICs to serve both service and storage network.
+
+We could take one solution from below to fit your requirement and security policy.
+
+### Disable DHCP, always setup Static IP on secondary NIC
+
+The issue is caused by race condition of two DHCP clients. We could disable DHCP and always setup static IP in your VM setting. Or you could only set secondary NIC to static IP and keep primary one with DHCP enable.
+
+1. edit `/etc/sysconfig/network/ifcfg-eth0` as primay NIC
+
+```
+BOOTPROTO='static'
+IPADDR='192.168.0.100'
+NETMASK='255.255.255.0'
+```
+
+Or setup `eth0` with DHCP
+
+```
+BOOTPROTO='dhcp'
+DHCLIENT_SET_DEFAULT_ROUTE='yes'
+```
+
+
+2. edit route table `/etc/sysconfig/network/ifroute-eth0` (skipping if you used DHCP in eth0)
+
+Do not put other default route for your secondary NIC
+
+```
+# Destination  Dummy/Gateway  Netmask  Interface
+default        192.168.0.254  -        eth0
+```
+
+3. setup secondary NIC, edit `/etc/sysconfig/network/ifcfg-eth1`
+
+```
+BOOTPROTO='static'
+IPADDR='10.0.0.100'
+NETMASK='255.255.255.0'
+```
+
+#### Cloud-Init config
+
+```yaml
+network:
+  version: 1
+  config:
+    - type: physical
+      name: enp1s0
+      subnets:
+        - type: dhcp
+    - type: physical
+      name: enp2s0
+      subnets:
+        - type: static
+          address: 10.0.0.100/24
+```
+   
+### Disable second NIC default route from DHCP
+
+If your NIC will need to get IP from DHCP, you should disable secondary NIC default route.
+
+1. edit `/etc/sysconfig/network/ifcfg-eth0` as primay NIC
+
+```
+BOOTPROTO='dhcp'
+DHCLIENT_SET_DEFAULT_ROUTE='yes'
+```
+
+2. setup secondary NIC, edit `/etc/sysconfig/network/ifcfg-eth1`
+
+```
+BOOTPROTO='dhcp'
+DHCLIENT_SET_DEFAULT_ROUTE='no'
+```
+
+#### Cloud-Init config
+
+This solution is not available in Cloud-Init. Cloud-Init didn't allow any option for DHCP.
```

Comment:
A cloud init section already exists above. Can we merge both?

Reply:
this cloud-init section is a sub-section of "Disable secondary NIC default route from DHCP".
It's not the same as the above one.
I want to mention there is no cloud-init config example and usage for "Disable secondary NIC default route from DHCP".